# Introduction

哈罗，各位 ~ 本文对[安泰杯 —— 跨境电商智能算法大赛](https://tianchi.aliyun.com/competition/entrance/231718/information)数据进行了初探与可视化处理，便于各位更为直观理解赛题数据与建模目标。 没有参加这个比赛的同学可能没有办法在天池这里下载到数据集，可以移步到这里下载：https://pan.baidu.com/s/18Z2mW8TFxwJvOWqmGXdPeg 提取码: iqrd

    
[赛题目标](https://tianchi.aliyun.com/competition/entrance/231718/introduction)：通过用户历史订单数据，预测用户下一次购买的商品。

[赛题数据](https://tianchi.aliyun.com/competition/entrance/231718/information)：第一轮初赛共四个文件：训练数据(Antai_AE_round1_train_20190626.csv)、测试数据(Antai_AE_round1_test_20190626.csv)、商品信息(Antai_AE_round1_item_attr_20190626.csv)、提交示例(Antai_AE_round1_submit_20190715.csv)

1. 训练数据：用户每次购买的商品id，订单日期以及用户国家标识

2. 测试数据：较于训练数据，测试数据剔除了用户需要预测最后一次购买记录

3. 商品信息：商品id、品类id、店铺id和商品价格

4. 提交示例：预测用户购买商品Top30的item_id依概率从高到低排序，buyer_admin_id,predict 1,predict 2,…,predict 30

第二轮复赛共三个文件：训练数据(Antai_AE_round2_train_20190813.csv)、测试数据(Antai_AE_round2_test_20190813.csv)、商品信息(Antai_AE_round2_item_attr_20190813.csv)

## 载入数据分析库包

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 
import gc
%matplotlib inline
pd.set_option('display.float_format', lambda x: '%.2f'%x)

## 读取数据

In [2]:
item = pd.read_csv('../data/Antai_AE/Antai_AE_round1_item_attr_20190626.csv')
train = pd.read_csv('../data/Antai_AE/Antai_AE_round1_train_20190626.csv')
test = pd.read_csv('../data/Antai_AE/Antai_AE_round1_test_20190626.csv')
submit = pd.read_csv('../data/Antai_AE/Antai_AE_round1_submit_20190715.csv', header=None)

In [3]:
# 商品属性表:
## item_id: 商品id
## cate: 类目id
## store_id: 店铺id
## item_price: 加密价格，其中价格的加密函数f(x)为一个单调增函数
item.head()

item_id  cate_id  store_id  item_price
0   240607     1495     12239           1
1   285232     1498     12239           1
2   113669     1503     12239           1
3   253601     1422     12239           1
4   246568     1497     12239           1

In [4]:
item.isnull().sum()

item_id       0
cate_id       0
store_id      0
item_price    0
dtype: int64

In [5]:
item["item_id"].nunique(), item["cate_id"].nunique(), item["store_id"].nunique(), item["item_price"].nunique()

(2832669, 4243, 95105, 20230)

In [6]:
# xx国的用户的购买数据和yy国的A部分用户的购买数据

# buyer_country_id: 买家国家id, 只有'xx'和'yy'两种取值
# buyer_admin_id: 买家id
# item_id: 商品id
# create_order_time: 订单创建时间
# irank: 每个买家对应的所有记录按照时间顺序的逆排序

train.head()

buyer_country_id  buyer_admin_id  item_id    create_order_time  irank
0               xx         8362078        1  2018-08-10 23:49:44     12
1               xx         9694304        2  2018-08-03 23:55:07      9
2               yy          101887        3  2018-08-27 08:31:26      3
3               xx         8131786        3  2018-08-31 06:00:19      9
4               xx         9778613        5  2018-08-21 06:01:56     14

In [7]:
train.isnull().sum()

buyer_country_id     0
buyer_admin_id       0
item_id              0
create_order_time    0
irank                0
dtype: int64

In [8]:
train.query("buyer_admin_id==8362078").sort_values(by="irank")

buyer_country_id  buyer_admin_id  item_id    create_order_time  irank
9725766               xx         8362078  9859480  2018-08-28 11:39:21      1
3487488               xx         8362078  3530082  2018-08-28 11:16:46      2
4080151               xx         8362078  4130245  2018-08-27 11:49:46      3
3305675               xx         8362078  3346056  2018-08-27 11:34:43      4
7182569               xx         8362078  7275850  2018-08-27 11:28:09      5
5795425               xx         8362078  5869190  2018-08-26 04:53:48      6
5795424               xx         8362078  5869190  2018-08-26 04:53:48      7
5768184               xx         8362078  5841575  2018-08-25 11:35:33      8
6866353               xx         8362078  6955075  2018-08-18 00:53:36      9
6866354               xx         8362078  6955075  2018-08-18 00:53:36     10
4406601               xx         8362078  4460989  2018-08-12 23:31:36     11
0                     xx         8362078        1  2018-08-10 23:49:44     12

In [9]:
# yy国的B部分用户的购买数据除掉最后一条
test.head()

buyer_country_id  buyer_admin_id   item_id    create_order_time  irank
0               yy         1061132    189045  2018-07-17 07:43:40     18
1               yy         1081430  10713670  2018-07-17 09:43:15      8
2               yy         1918135  10546322  2018-07-17 11:11:34      9
3               yy          103277   5199603  2018-07-17 13:27:44     24
4               yy          103277   2477273  2018-07-17 13:27:44     22

In [10]:
test.isnull().sum()

buyer_country_id     0
buyer_admin_id       0
item_id              0
create_order_time    0
irank                0
dtype: int64

In [11]:
# 数据集特点：
# 1）每个用户有至少7条购买数据；
# 2）测试数据中每个用户的最后一条购买数据(去掉了)所对应的商品一定在训练数据中出现过；
# 3）少量用户在两个国家有购买记录，评测中将忽略这部分记录；
train.groupby("buyer_admin_id").size().min(), test.groupby("buyer_admin_id").size().min()

(8, 7)

In [12]:
pd.concat([train, test]).groupby("buyer_admin_id")["buyer_country_id"].nunique().value_counts()

1    820503
2       102
Name: buyer_country_id, dtype: int64

In [13]:
two_country_buyers = pd.concat([train, test]).groupby("buyer_admin_id")["buyer_country_id"].nunique()\
    .reset_index().query("buyer_country_id==2")["buyer_admin_id"].tolist()

train_buyers = train["buyer_admin_id"].unique().tolist()
for b in two_country_buyers:
    if b not in train_buyers:
        print(b)

In [14]:
submit.head()

0         1        2        3        4         5         6         7   \
0  152    343131  2140290  2600629  1554159  12211445    776612   1630164   
1  282   8988184  2210260    96479  4052816   7982989  11143899  10070236   
2  321  11140901  4618502  9666424   421308   8443514  12341414   4133929   
3  809    692592  9851186  5206030  1210732   6016462   6971778  11615323   
4  870   8361014  2327622  5481981  7530272  11680535   6555249   6137574   

        8         9   ...        21        22        23       24       25  \
0  3901301   8353580  ...   2327622   2014295   5076158  5042708   818444   
1  2002400   8757854  ...   5630838   8607737   7740462  7028968  9167824   
2  4374880   9104983  ...  12975999  10793761   7524523   666863   204808   
3  7263674  10901782  ...   3227122  10539551   2802000  8251066  9515496   
4   914424  10445923  ...   8757854  10578287  10715090  7176446  3165153   

         26        27       28        29       30  
0  10451831   4333367  9437487   8469786  9465790  
1  12938507   6399956  3109441  10636176  6255625  
2   8893340   7431033  1773069  12970909  9373627  
3   7065316  11997137  4438427   6567724  7212577  
4  10144005   8753413  2800387  12152573  1068526  

[5 rows x 31 columns]

In [15]:
submit.shape, test["buyer_admin_id"].nunique()

((11398, 31), 11398)

### 数据预处理
* 合并train和test文件
* 提取日期年月日等信息
* 关联商品价格、品类、店铺
* 转化每列数据类型为可存储的最小值，减少内存消耗
* 保存为hdf5格式文件，加速读取

In [16]:
df = pd.concat([train.assign(is_train=1), test.assign(is_train=0)])

df['create_order_time'] = pd.to_datetime(df['create_order_time'])
df['date'] = df['create_order_time'].dt.date
df['day'] = df['create_order_time'].dt.day
df['hour'] = df['create_order_time'].dt.hour

df = pd.merge(df, item, how='left', on='item_id')

In [17]:
df.head()

buyer_country_id  buyer_admin_id  item_id   create_order_time  irank  \
0               xx         8362078        1 2018-08-10 23:49:44     12   
1               xx         9694304        2 2018-08-03 23:55:07      9   
2               yy          101887        3 2018-08-27 08:31:26      3   
3               xx         8131786        3 2018-08-31 06:00:19      9   
4               xx         9778613        5 2018-08-21 06:01:56     14   

   is_train        date  day  hour  cate_id  store_id  item_price  
0         1  2018-08-10   10    23  2324.00  10013.00     4501.00  
1         1  2018-08-03    3    23  3882.00   4485.00     2751.00  
2         1  2018-08-27   27     8   155.00   8341.00      656.00  
3         1  2018-08-31   31     6   155.00   8341.00      656.00  
4         1  2018-08-21   21     6  1191.00   1949.00     1689.00

In [18]:
memory = df.memory_usage().sum() / 1024**2 
print('Before memory usage of properties dataframe is :', memory, " MB")

dtype_dict = {'buyer_admin_id' : 'int32', 
              'item_id' : 'int32', 
              'store_id' : pd.Int32Dtype(),
              'irank' : 'int16',
              'item_price' : pd.Int16Dtype(),
              'cate_id' : pd.Int16Dtype(),
              'is_train' : 'int8',
              'day' : 'int8',
              'hour' : 'int8',
             }

df = df.astype(dtype_dict)
memory = df.memory_usage().sum() / 1024**2 
print('After memory usage of properties dataframe is :', memory, " MB")
del train, test; 
# gc.collect()

Before memory usage of properties dataframe is : 1292.8728713989258  MB
After memory usage of properties dataframe is : 696.1623153686523  MB


In [21]:
df.isnull().any()

buyer_country_id     False
buyer_admin_id       False
item_id              False
create_order_time    False
irank                False
is_train             False
date                 False
day                  False
hour                 False
cate_id               True
store_id              True
item_price            True
dtype: bool

In [22]:
df[['store_id', 'item_price', 'cate_id']].min()

store_id     1.00
item_price   1.00
cate_id      1.00
dtype: float64

In [23]:
for col in ['store_id', 'item_price', 'cate_id']:
    df[col] = df[col].fillna(0).astype(np.int32).replace(0, np.nan)
df.to_hdf('../data/train_test.h5', '1.0')

/Users/mac/.local/share/virtualenvs/AntaiCup-vTCStMaD/lib/python3.9/site-packages/tables/path.py:137: NaturalNameWarning: object name is not a valid Python identifier: '1.0'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
  check_attribute_name(name)
/var/folders/ls/skqby2m95l3_qvzb_rc2d0q40000gn/T/ipykernel_72676/251884372.py:3: PerformanceWarning: 
your performance may suffer as PyTables will pickle object types that it cannot
map directly to c-types [inferred_type->mixed,key->block5_values] [items->Index(['buyer_country_id', 'date'], dtype='object')]

  df.to_hdf('../data/train_test.h5', '1.0')


In [24]:
%%time
df = pd.read_hdf('../data/train_test.h5', '1.0')

CPU times: user 4.52 s, sys: 1.96 s, total: 6.48 s
Wall time: 6.85 s


In [26]:
%%time
train = pd.read_csv('../data/Antai_AE/Antai_AE_round1_train_20190626.csv')
test = pd.read_csv('../data/Antai_AE/Antai_AE_round1_test_20190626.csv')
item = pd.read_csv('../data/Antai_AE/Antai_AE_round1_item_attr_20190626.csv')
del train, test;

CPU times: user 8.57 s, sys: 1.11 s, total: 9.67 s
Wall time: 9.79 s


# Overview: 数据内容

In [27]:
df.head()

buyer_country_id  buyer_admin_id  item_id   create_order_time  irank  \
0               xx         8362078        1 2018-08-10 23:49:44     12   
1               xx         9694304        2 2018-08-03 23:55:07      9   
2               yy          101887        3 2018-08-27 08:31:26      3   
3               xx         8131786        3 2018-08-31 06:00:19      9   
4               xx         9778613        5 2018-08-21 06:01:56     14   

   is_train        date  day  hour  cate_id  store_id  item_price  
0         1  2018-08-10   10    23  2324.00  10013.00     4501.00  
1         1  2018-08-03    3    23  3882.00   4485.00     2751.00  
2         1  2018-08-27   27     8   155.00   8341.00      656.00  
3         1  2018-08-31   31     6   155.00   8341.00      656.00  
4         1  2018-08-21   21     6  1191.00   1949.00     1689.00

In [ ]:
# Null 空值统计
for pdf in [df, item]:
    for col in pdf.columns:
        print(col, pdf[col].isnull().sum())

In [ ]:
df.describe()

In [ ]:
item.describe()

数据内容：
* 用户、商品、店铺、品类乃至商品价格都是从1开始用整数编号
* 订单日期格式为：YYYY-mm-dd HH:mm:ss
* 源数据中都木有空值，但是由于某些商品，不在商品表，因此缺少了一些价格、品类信息。

# 数据探查

下一步，我们依次对每个文件的特征进行基础统计和可视化处理，这是对数据进一步理解的基础。

[]~(￣▽￣)~* Let's do it.

## 训练集与测试集

In [ ]:
train = df['is_train']==1
test = df['is_train']==0

In [ ]:
train_count = len(df[train])
print('训练集样本量是',train_count)
test_count = len(df[test])
print('测试集样本量是',test_count)
print('样本比例为：', train_count/test_count)

### buyer_country_id 国家编号

In [ ]:
def groupby_cnt_ratio(df, col):
    if isinstance(col, str):
        col = [col]
    key = ['is_train', 'buyer_country_id'] + col
    
    # groupby function
    cnt_stat = df.groupby(key).size().to_frame('count')
    ratio_stat = (cnt_stat / cnt_stat.groupby(['is_train', 'buyer_country_id']).sum()).rename(columns={'count':'count_ratio'})
    return pd.merge(cnt_stat, ratio_stat, on=key, how='outer').sort_values(by=['count'], ascending=False)

In [ ]:
groupby_cnt_ratio(df, [])

In [ ]:
plt.figure(figsize=(8,6))
sns.countplot(x='is_train', data = df, palette=['red', 'blue'], hue='buyer_country_id', order=[1, 0])
plt.xticks(np.arange(2), ('训练集', '测试集'))
plt.xlabel('数据文件')
plt.title('国家编号');

buyer_country_id 国家编号

> 本次比赛给出若干日内来自成熟国家的部分用户的行为数据，以及来自待成熟国家的A部分用户的行为数据，以及待成熟国家的B部分用户的行为数据去除每个用户的最后一条购买数据，让参赛人预测B部分用户的最后一条行为数据。

* 训练集中有2个国家数据，xx国家样本数10635642，占比83%，yy国家样本数2232867条，仅占17%
* 预测集中有yy国家的166832数据, 训练集中yy国样本数量是测试集中的13倍，如赛题目的所言，期望通过大量成熟国家来预测少量带成熟国家的用户购买行为

### buyer_admin_id 用户编号

In [ ]:
print('训练集中用户数量',len(df[train]['buyer_admin_id'].unique()))
print('测试集中用户数量',len(df[test]['buyer_admin_id'].unique()))

In [ ]:
union = list(set(df[train]['buyer_admin_id'].unique()).intersection(set(df[test]['buyer_admin_id'].unique())))
print('同时在训练集测试集出现的有6位用户，id如下：',union)

In [ ]:
df[train][df['buyer_admin_id'].isin(union)].sort_values(by=['buyer_admin_id','irank']).head(10)

In [ ]:
df[test][df['buyer_admin_id'].isin(union)].sort_values(by=['buyer_admin_id','irank']).head(3)

In [ ]:
df[(train) & (df['irank']==1) & (df['buyer_admin_id'].isin(['12858772','3106927','12368445']))]

emmm... 为啥同一个用户在训练集和测试集国家不一样了呢？但是其他信息能对上。。。，而且rank=1的结果直接给出来了。。。

id为12858772、3106927、12368445直接把结果给出来

可能是数据清洗出问题了，后面再看看怎么处理

#### 用户记录数分布

In [ ]:
admin_cnt = groupby_cnt_ratio(df, 'buyer_admin_id')
admin_cnt.groupby(['is_train','buyer_country_id']).head(3)

In [ ]:
# 用户购买记录数——最多、最少、中位数
admin_cnt.groupby(['is_train','buyer_country_id'])['count'].agg(['max','min','median'])

In [ ]:
fig, ax = plt.subplots(1, 2 ,figsize=(16,6))
ax[0].set(xlabel='用户记录数')
sns.kdeplot(admin_cnt.loc[(1, 'xx')]['count'].values, ax=ax[0]).set_title('训练集--xx国用户记录数')

ax[1].legend(labels=['训练集', '测试集'], loc="upper right")
ax[1].set(xlabel='用户记录数')
sns.kdeplot(admin_cnt[admin_cnt['count']<50].loc[(1, 'yy')]['count'].values, ax=ax[1]).set_title('yy国用户记录数')
sns.kdeplot(admin_cnt[admin_cnt['count']<50].loc[(0, 'yy')]['count'].values, ax=ax[1]);

用户记录数进行了一波简单的探查：
* 训练集中记录了*809213*个用户的数据，其中id为10828801的用户拔得头筹，有42751条购买记录，用户至少都有8条记录
* 训练集中记录了*11398*个用户的数据，其中id为2041038的用户勇冠三军，有1386条购买记录，用户至少有7条记录

Notes: 验证集中用户最少仅有7条，是因为最后一条记录被抹去

从上面数据和图表看到，用户记录数大都都分布在0~50，少量用户记录甚至超过了10000条，下一步对用户记录数分布继续探索

In [ ]:
admin_cnt.columns = ['记录数', '占比']
admin_user_cnt = groupby_cnt_ratio(admin_cnt, '记录数')
admin_user_cnt.columns = ['人数', '人数占比']
admin_user_cnt.head()

In [ ]:
# xx国——用户记录数与用户数
admin_user_cnt.loc[(1,'xx')][['人数','人数占比']].T

In [ ]:
# yy国——记录数与用户数占比
admin_user_cnt.loc[([1,0],'yy',slice(None))][['人数','人数占比']].unstack(0).drop('人数',1).head(10)

In [ ]:
fig, ax = plt.subplots(2, 1, figsize=(16,10))
admin_plot = admin_user_cnt.reset_index()
sns.barplot(x='记录数', y='人数占比', data=admin_plot[(admin_plot['记录数']<50) & (admin_plot['buyer_country_id']=='xx')], 
            estimator=np.mean, ax=ax[0]).set_title('训练集——xx国记录数与人数占比');

sns.barplot(x='记录数', y='人数占比', hue='is_train', data=admin_plot[(admin_plot['记录数']<50) & (admin_plot['buyer_country_id']=='yy')], 
            estimator=np.mean, ax=ax[1]).set_title('yy国记录数与人数占比');

用户记录数进一步探查结论：
    * 不管是训练集还是验证集，99%的用户购买记录都在50条内，这是比较符合正常逻辑
    * TODO:对于发生大量购买行为的用户，后面再单独探查，是否有其他规律或疑似刷单现象

### item_id 商品编号

In [ ]:
print('商品表中商品数：',len(item['item_id'].unique()))
print('训练集中商品数：',len(df[train]['item_id'].unique()))
print('验证集中商品数：',len(df[test]['item_id'].unique()))
print('仅训练集有的商品数：',len(list(set(df[train]['item_id'].unique()).difference(set(df[test]['item_id'].unique())))))
print('仅验证集有的商品数：',len(list(set(df[test]['item_id'].unique()).difference(set(df[train]['item_id'].unique())))))
print('训练集验证集共同商品数：',len(list(set(df[train]['item_id'].unique()).intersection(set(df[test]['item_id'].unique())))))
print('训练集中不在商品表的商品数：',len(list(set(df[train]['item_id'].unique()).difference(set(item['item_id'].unique())))))
print('验证集中不在商品表的商品数：',len(list(set(df[test]['item_id'].unique()).difference(set(item['item_id'].unique())))))

#### 商品销量

In [ ]:
item_cnt = groupby_cnt_ratio(df, 'item_id')
item_cnt.columns=['销量', '总销量占比']
item_cnt.reset_index(inplace=True)

In [ ]:
top_item_plot = item_cnt.groupby(['is_train','buyer_country_id']).head(10)

In [ ]:
fig, ax = plt.subplots(2, 1, figsize=(16,12))
sns.barplot(x='item_id', y='销量', data=top_item_plot[top_item_plot['buyer_country_id']=='xx'], 
            order=top_item_plot['item_id'][top_item_plot['buyer_country_id']=='xx'], ax=ax[0], estimator=np.mean).set_title('xx国-TOP热销商品')
sns.barplot(x='item_id', y='销量', hue='is_train', data=top_item_plot[top_item_plot['buyer_country_id']=='yy'], 
            order=top_item_plot['item_id'][top_item_plot['buyer_country_id']=='yy'], ax=ax[1], estimator=np.mean).set_title('yy国-TOP热销商品');

初步数据发现：
* 训练集中出售最多商品是12691565，卖了112659次。
* 训练集中出售最多商品是5595070，卖了112659次。
* 大部分商品只有1次出售记录，符合电商长尾属性
* 比较奇怪的yy国中，训练集和测试集中热销商品并不太一样

#### 整体商品销量分布

In [ ]:
item_order_cnt = groupby_cnt_ratio(item_cnt, '销量')
item_order_cnt.columns = ['商品数', '占比']

In [ ]:
item_order_cnt.groupby(['is_train','buyer_country_id']).head(5).sort_values(by=['buyer_country_id','is_train'])

In [ ]:
item_order_plot = item_order_cnt.reset_index()
item_order_plot = item_order_plot[item_order_plot['销量']<=8]

xx_item_order_plot = item_order_plot[item_order_plot['buyer_country_id']=='xx']
yy_item_order_plot = item_order_plot[item_order_plot['buyer_country_id']=='yy']
yy_item_order_plot_1 = yy_item_order_plot[yy_item_order_plot['is_train']==1]
yy_item_order_plot_0 = yy_item_order_plot[yy_item_order_plot['is_train']==0]

In [ ]:
# 商品销量饼图
def text_style_func(pct, allvals):
    absolute = int(pct/100.*np.sum(allvals))
    return "{:.1f}%({:d})".format(pct, absolute)

def pie_param(ax, df, color_palette):
    return ax.pie(df['占比'].values, autopct=lambda pct: text_style_func(pct, df['商品数']), labels = df['销量'], 
                  explode = [0.1]+ np.zeros(len(df)-1).tolist(), pctdistance = 0.7, colors=sns.color_palette(color_palette, 8))

fig, ax = plt.subplots(1, 3, figsize=(16,12))
ax[0].set(xlabel='xx国-商品销量')
ax[0].set(ylabel='xx国-商品数量比例')
pie_param(ax[0], xx_item_order_plot, "coolwarm")
ax[1].set(xlabel='yy国-训练集商品销量')
pie_param(ax[1], yy_item_order_plot_1, "Set3")
ax[2].set(xlabel='yy国测试集集商品销量')
pie_param(ax[2], yy_item_order_plot_0, "Set3");

In [ ]:
print(xx_item_order_plot.head(10)['占比'].sum())
print(yy_item_order_plot_1.head(10)['占比'].sum())
print(yy_item_order_plot_0.head(10)['占比'].sum())

总体来看，由于训练集数据远多于测试集数据：
* 训练集商品销量大于测试集商品销量
* 长尾趋势严重，热门商品少，大量商品仅有数次销售记录，1单商品占了绝大部分(均超过50%)
* 训练集中92%的商品销量不超过10件，而在测试集中97%的商品销量不超过10件
* 此外训练集中yy国的商品销量略大于测试集

### cate_id 品类编号

In [ ]:
print('商品品类数', len(item['cate_id'].unique()))
print('训练集商品品类数', len(df[train]['cate_id'].unique()))
print('测试集商品品类数', len(df[test]['cate_id'].unique()))

#### 各个品类下商品数量

In [ ]:
cate_cnt = item.groupby(['cate_id']).size().to_frame('count').reset_index()
cate_cnt.sort_values(by=['count'], ascending=False).head(5)

In [ ]:
plt.figure(figsize=(12,4))
sns.kdeplot(data=cate_cnt[cate_cnt['count']<1000]['count']);

我们发现：
    * 579品类一花独秀有17W个商品，可能是平台主营方向
    * 大部分品类都在100个以上

### store_id 店铺编号

In [ ]:
print('商品店铺数', len(item['store_id'].unique()))
print('训练集店铺数', len(df[train]['store_id'].unique()))
print('测试集店铺数', len(df[train]['store_id'].unique()))

#### 店铺下品类数量

In [ ]:
store_cate_cnt = item.groupby(['store_id'])['cate_id'].nunique().to_frame('count').reset_index()
store_cate_cnt.sort_values(by=['count'], ascending=False).head(5)

In [ ]:
store_cnt_cate_cnt = store_cate_cnt.groupby(['count']).size().reset_index()
store_cnt_cate_cnt.columns = ['店铺品类数', '店铺数量']

In [ ]:
plt.figure(figsize=(12,4))
sns.barplot(x='店铺品类数', y='店铺数量', data=store_cnt_cate_cnt[store_cnt_cate_cnt['店铺品类数']<50], estimator=np.mean);

#### 店铺下商品数量

In [ ]:
store_item_cnt = item.groupby(['store_id'])['item_id'].nunique().to_frame('count').reset_index()
store_item_cnt.sort_values(by=['count'], ascending=False).head(5)

In [ ]:
store_cnt_item_cnt = store_item_cnt.groupby(['count']).size().reset_index()
store_cnt_item_cnt.columns = ['店铺商品数', '店铺数量']

In [ ]:
store_cnt_item_cnt.T

In [ ]:
plt.figure(figsize=(16,4))
sns.barplot(x='店铺商品数', y='店铺数量', data=store_cnt_item_cnt[store_cnt_item_cnt['店铺商品数']<80], estimator=np.mean);

#### item_price 商品价格

In [ ]:
print(item['item_price'].max(), item['item_price'].min(), item['item_price'].mean(), item['item_price'].median())

In [ ]:
plt.figure(figsize=(16,4))
plt.subplot(121)
sns.kdeplot(item['item_price'])
plt.subplot(122)
sns.kdeplot(item['item_price'][item['item_price']<1000]);

In [ ]:
price_cnt = item.groupby(['item_price']).size().to_frame('count').reset_index()
price_cnt.sort_values(by=['count'], ascending=False).head(10)

关于商品价格：商品价格是通过函数转化成了从1开始的整数，最大值为20230，最小值为1。
    * 经常对商品价格统计，大部门商品都是整百数，Top5价格200\500\100\400\300
    * TODO：整百商品探查

#### 有售商品价格

In [ ]:
print(df[train]['item_price'].max(), df[train]['item_price'].min(), df[train]['item_price'].mean(), df[train]['item_price'].median())
print(df[test]['item_price'].max(), df[test]['item_price'].min(), df[test]['item_price'].mean(), df[test]['item_price'].median())

In [ ]:
plt.figure(figsize=(12,4))
sns.kdeplot(df[train][df[train]['item_price']<1000][['item_id','item_price']].drop_duplicates()['item_price'])
sns.kdeplot(df[test][df[test]['item_price']<1000][['item_id','item_price']].drop_duplicates()['item_price']);

商品价格与销量

In [ ]:
df[train].groupby(['item_price'])['item_id'].nunique().to_frame('商品数量').head()

In [ ]:
price_cnt = groupby_cnt_ratio(df, 'item_price')
price_cnt.groupby(['is_train', 'buyer_country_id']).head(5)

似乎价格与销量并无直接关系
    * 但是价格为100、200、300、400、500整百数位居销量榜
    * xx国，17844如此高价格的商品销量这么高？

### create_order_time 订单日期

In [ ]:
print(df[train]['create_order_time'].min(), df[train]['create_order_time'].max())
print(df[test]['create_order_time'].min(), df[test]['create_order_time'].max())

In [ ]:
train_df_seven = df[train][df[train]['create_order_time']<pd.to_datetime('2018-08-01')]
train_df_eight = df[train][df[train]['create_order_time']>pd.to_datetime('2018-08-01')]
train_df_seven = df[train][df[train]['create_order_time']<pd.to_datetime('2018-08-01')]
train_df_eight = df[train][df[train]['create_order_time']>pd.to_datetime('2018-08-01')]

In [ ]:
print('7月数据量',len(df[train][df[train]['create_order_time']<pd.to_datetime('2018-08-01')]),
      '\n8月数据量',len(df[train][df[train]['create_order_time']>pd.to_datetime('2018-08-02')]))

In [ ]:
date_cnt = groupby_cnt_ratio(df, 'date')
date_cnt.columns = ['当天销量', "占比"]
date_cnt = date_cnt.reset_index()

In [ ]:
fig, ax = plt.subplots(2, 1, figsize=(16,10))
sns.lineplot(x='date', y='当天销量', hue='buyer_country_id', data=date_cnt[(date_cnt['is_train']==1)], 
            estimator=np.mean, ax=ax[0]).set_title('训练集——每日销量');

sns.lineplot(x='date', y='当天销量', hue='is_train', data=date_cnt[(date_cnt['buyer_country_id']=='yy')], 
            estimator=np.mean, ax=ax[1]).set_title('yy国每日销量');

很明显：
* 训练集中7月份数据远小于8月份数据
* 训练集中xx国和yy国每日销量趋势十分相似，且在27日有个波峰

In [ ]:
seven = date_cnt[date_cnt['date']<pd.to_datetime('2018-08-02')]
eight = date_cnt[date_cnt['date']>=pd.to_datetime('2018-08-02')]

In [ ]:
fig, ax = plt.subplots(2, 3, figsize=(20,16))
def barplot(ax, df, title):
    df['date'] = df['date'].astype(str)
    sns.barplot(y='date', x='当天销量' ,data=df, order=sorted(df['date'].unique()), ax=ax, estimator=np.mean)\
    .set_title(title)
    
barplot(ax[0][0], seven[(seven['is_train']==1) & (seven['buyer_country_id']=='xx')], 'xx国7月份销量')
barplot(ax[1][0], eight[(eight['is_train']==1) & (eight['buyer_country_id']=='xx')], 'xx国8月份销量')
barplot(ax[0][1], seven[(seven['is_train']==1) & (seven['buyer_country_id']=='yy')], '训练集-yy国7月份销量')
barplot(ax[1][1], eight[(eight['is_train']==1) & (eight['buyer_country_id']=='yy')], '训练集-yy国8月份销量')
barplot(ax[0][2], seven[(seven['is_train']==0) & (seven['buyer_country_id']=='yy')], '测试集-yy国7月份销量')
barplot(ax[1][2], eight[(eight['is_train']==0) & (eight['buyer_country_id']=='yy')], '测试集-yy国8月份销量')
plt.tight_layout()

数据放大后看：
* 训练集和测试集在8月份有相似的波动规律，27号出现波峰，当天剧增数据有待下一步探查

#### 每日uv与商品数(去重)

In [ ]:
unique = df.groupby(['is_train', 'buyer_country_id', 'date']).agg({'buyer_admin_id':'nunique','item_id':['nunique','size']})
unique.columns = ['uv','商品数(去重)', '销量']
unique = unique.reset_index()
unique = pd.melt(unique, id_vars=['is_train', 'buyer_country_id', 'date'], value_vars=['uv', '商品数(去重)', '销量'])
unique['date'] = unique['date'].astype(str)
unique = unique[unique['date']>='2018-08-02']

In [ ]:
fig, ax = plt.subplots(3, 1, figsize=(16,8), sharex=True)
sns.lineplot(x='date', y='value', hue='variable', data=unique[(unique['is_train']==1) & (unique['buyer_country_id']=='xx')], 
             estimator=np.mean, ax=ax[0]).set_title('xx国每日销售数据');

sns.lineplot(x='date', y='value', hue='variable', data=unique[(unique['is_train']==0) & (unique['buyer_country_id']=='yy')], 
            estimator=np.mean, ax=ax[1]).set_title('训练集-yy国每日销量');

sns.lineplot(x='date', y='value', hue='variable', data=unique[(unique['is_train']==1) & (unique['buyer_country_id']=='yy')], 
            estimator=np.mean, ax=ax[2]).set_title('测试集-yy国每日销量')
plt.xticks(rotation=90);

对每日的uv、商品数和销量作图发现：
* 三者基本上呈正相关，xx国的商品单品销量更高

# BASELINE
选取用户近30次购买记录作为预测值，越近购买的商品放在越靠前的列，不够30次购买记录的用热销商品5595070填充

In [ ]:
test = pd.read_csv('../data/Antai_AE_round1_test_20190626.csv')
tmp = test[test['irank']<=31].sort_values(by=['buyer_country_id', 'buyer_admin_id', 'irank'])[['buyer_admin_id','item_id','irank']]
sub = tmp.set_index(['buyer_admin_id', 'irank']).unstack(-1)
sub.fillna(5595070).astype(int).reset_index().to_csv('../submit/sub.csv', index=False, header=None)

In [ ]:
sub = pd.read_csv('../submit/sub.csv', header = None)
sub.head()